data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterTest.java

```diff
@@ -138,11 +153,85 @@ void route_with_multiple_DataFlowComponents() {
+            Record record4 = mock(Record.class);
+            when(record3.getData()).thenReturn(event3);
+            Object notAnEvent = mock(Object.class);
+            when(record4.getData()).thenReturn(notAnEvent);
+            List<Record> recordsIn = List.of(record1, record2, record3, record4);
+            Map<Record, Set<String>> recordsToRoutes = new HashMap<>();
+            recordsToRoutes.put(record1, Set.of(UUID.randomUUID().toString()));
+            recordsToRoutes.put(record2, Set.of(UUID.randomUUID().toString()));
+            recordsToRoutes.put(record3, Set.of());
+            recordsToRoutes.put(record4, Set.of());
+            when(routeEventEvaluator.evaluateEventRoutes(recordsIn)).thenReturn(recordsToRoutes);
+            dataFlowComponents = new ArrayList<>();
+            for (int i = 0; i < 5; i++) {
+                final DataFlowComponent dataFlowComponent = mock(DataFlowComponent.class);
+                dataFlowComponents.add(dataFlowComponent);
+            }
+
+            createObjectUnderTest().route(recordsIn, dataFlowComponents, getRecordStrategy, componentRecordsConsumer);
+
+            for (DataFlowComponent<TestComponent> dataFlowComponent : dataFlowComponents) {
+                verify(dataFlowComponentRouter).route(eq(recordsIn), eq(dataFlowComponent), eq(recordsToRoutes), eq(getRecordStrategy), any(BiConsumer.class));
+            }
+            // Verify noRouteHandler gets invoked only for record3 and not
+            // for record4, because record4 has non-Event type data
+            verify(noRouteHandler, times(1)).accept(any());
+        }
+
+        @Test
+        void route_with_multiple_DataFlowComponents_with_unrouted_events_and_sink_with_noroutes() {
+            Event event1 = mock(Event.class);
+            Event event2 = mock(Event.class);
+            Event event3 = mock(Event.class);
+            EventHandle eventHandle3 = mock(EventHandle.class);
+            Record record1 = mock(Record.class);
+            Record record2 = mock(Record.class);
+            Record record3 = mock(Record.class);
+            Record record4 = mock(Record.class);
+            Object notAnEvent = mock(Object.class);
+            List<Record> recordsIn = List.of(record1, record2, record3, record4);
+            Map<Record, Set<String>> recordsToRoutes = new HashMap<>();
+            recordsToRoutes.put(record1, Set.of(UUID.randomUUID().toString()));
+            recordsToRoutes.put(record2, Set.of(UUID.randomUUID().toString()));
+            recordsToRoutes.put(record3, Set.of());
+            recordsToRoutes.put(record4, Set.of());
+            when(routeEventEvaluator.evaluateEventRoutes(recordsIn)).thenReturn(recordsToRoutes);
+            dataFlowComponents = new ArrayList<>();
+            for (int i = 0; i < 5; i++) {
+                final DataFlowComponent dataFlowComponent = mock(DataFlowComponent.class);
+                dataFlowComponents.add(dataFlowComponent);
+                if (i == 0) {
```

Comment:
It might be cleaner to move the mocking out of the condition and only have the set in the condition.


```
final Set<String> routes;
if(i == 0) {
  routes = Collections.emptySet();
}
else {
  routes = Set.of("aa");
}
when(dataFlowComponent.getRoutes()).thenReturn(routes);
```

or

```
final Set<String> routes = i ==0 ? Collections.emptySet() : Set.of("aa");
when(dataFlowComponent.getRoutes()).thenReturn(routes);
```